src/NLog/Config/ServiceRepository.cs

```diff
@@ -79,32 +82,97 @@ public void RegisterType(Type type, ConfigurationItemCreator objectResolver)
                 _serviceRepository[type] = objectResolver;
         }
 
-        public object ResolveInstance(Type itemType)
+        public object ResolveService(Type itemType)
         {
             var createInstance = CreateInstance;
             if (createInstance != null)
             {
                 return createInstance(itemType);
             }
 
-            return DefaultResolveInstance(itemType);
+            return DefaultResolveInstance(itemType, new HashSet<Type>());
         }
 
-        private object DefaultResolveInstance(Type itemType)
+        private object DefaultResolveInstance(Type itemType, HashSet<Type> seenTypes)
         {
+            InternalLogger.Trace("Resolve {0}", itemType.FullName);
             if (_serviceRepository.TryGetValue(itemType, out var objectResolver))
             {
+                InternalLogger.Trace("Resolve {0} done", itemType.FullName);
                 return objectResolver(itemType);
             }
 
             try
             {
-                return Activator.CreateInstance(itemType);
+                //todo DI cache/find upfront
+                var defaultConstructor = itemType.GetConstructor(Type.EmptyTypes);
```

Comment:
Can this lookup be cached?

Reply:
Yes, it's a TODO ;)

Maybe another PR?

Reply:
Sorry cannot do dependency injection without proper performance :)

Reply:
This isn't to the master /dev yet.

;)

Reply:
we could add it after this. I could do it, or maybe you if you like.

Reply:
trying to keeping the PRS small

Reply:
PS with the default ctor is as fast as before ;)

Reply:
No this is your baby :)

Reply:
For performance we need much more, e.g. another Register so we could do the reflection stuff on registering. :)

Reply:
Not talking about the performance of calling. It was the `itemType.GetConstructor`-lookup. `Activator.CreateInstance` performs caching. You need to re-implement that here instead.

Reply:
OK, so I should prefer Activator.CreateInstance here? 

> Not talking about the performance of calling actual constructor. It was the itemType.GetConstructor-lookup

I'm also not. You could do the lookup on register.

Reply:
No when you have performed the constructor-lookup, then it doesn't make sense to ask `Activator.CreateInstance` to also perform constructor-lookup.

Reply:
OK, will do in another PR

PS: We need at the end a final PR to dev, then we can soon/refactor also.

Reply:
I'm still not sure why we have this feature-branch. Not like dev is released tomorrow. But I will wait until you are done. Then I will look into the LogEventInfo-part.

Reply:
Because it pollutes dev and I try to keep PRs small